### DISCUSSION TITLE
Lambda statement uses incorrect cached values for `BinaryExpression` after engine disposed and recreated

### DISCUSSION THREAD

**Discussion by provinzkraut** [sqlalchemy/sqlalchemy#12083] [Usage Questions]
This is a bit long winded but essentially, so I'll start with the reproducer:

```python
from sqlalchemy import lambda_stmt, select, create_engine
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, Session


class Base(DeclarativeBase):
    pass


class SomeModel(Base):
    __tablename__ = "some"
    id: Mapped[int] = mapped_column(primary_key=True)
    foo: Mapped[str]
    bar: Mapped[str]


def append_to_lambda(stmt, crt):
    stmt += lambda s: s.where(crt)
    return stmt


def make_query(stmt, *criteria):
    for crt in criteria:
        stmt = append_to_lambda(stmt, crt)

    return stmt


def main():
    engine = create_engine("sqlite+pysqlite://")
    Base.metadata.create_all(bind=engine)

    with Session(bind=engine) as session:
        session.add_all(
            [
                SomeModel(foo="one", bar="one"),
                SomeModel(foo="two", bar="two"),
            ]
        )
        session.commit()

        stmt = lambda_stmt(lambda: select(SomeModel))
        # create a filter criterion that will never match anything
        stmt = make_query(
            stmt,
            SomeModel.foo == "one",
            SomeModel.bar == "two",
        )

        assert len(session.scalars(stmt).all()) == 0

    engine.dispose()
```

Calling `main` once works as expected. However, when calling `main` two times consecutively, it will match a row, despite no such row being present in the database.

Note that it's necessary to recreate the engine for this to occur; Moving the engine declaration outside `main` will make it behave as expected. Another factor is creating the lambda inside a separate function, which I guess makes sense since this is probably related to some closure behaviour.

Turning on `echo` gives us an insight into what's happening. The second `SELECT` prints out:
```
...
2024-11-12 19:33:13,887 INFO sqlalchemy.engine.Engine SELECT some.id, some.foo, some.bar 
FROM some 
WHERE some.foo = ? AND some.bar = ?
2024-11-12 19:33:13,887 INFO sqlalchemy.engine.Engine [generated in 0.00014s] ('two', 'two')
```

So it is generating a statement that was never there.

I've played around with it a bit, and found this trick to make it "work":

```python
def make_query(stmt, *criteria):
    for crt in criteria:
        stmt = stmt.add_criteria(lambda s: s.where(crt), track_on=[id(crt)])

    return stmt
```

Which isn't surprising, since would effectively be a unique value every time.

<hr>

I'm not sure what the correct behaviour is here, but this one is certainly not what I suspected, and I'm not entirely sure if this is a user error or a genuine bug :)

<hr>

*Updated example to fix an issue: https://github.com/sqlalchemy/sqlalchemy/discussions/12083#discussioncomment-11230973*

**Comment by zzzeek** [comment]
sure looks like a bug but I would say to just avoid the lambda statement concept, which proved to be way too weird to be reliable.   I likely should just deprecate it since these issues are impossible to continue tracking down.

**Reply by zzzeek** [reply]
though this one is really strange in that it's local to the new engine being created.     that's very odd

**Reply by CaselIT** [reply]
it's just python being wired here. inline function do not make a closure, so crt is always the last one

**Reply by provinzkraut** [reply]
> I would say to just avoid the lambda statement concept, which proved to be way too weird to be reliable. I likely should just deprecate it since these issues are impossible to continue tracking down.

Could you elaborate on this, especially the part about the deprecation? We have quite a bit of code using lambdas at the moment and are currently debating whether to keep it in or not, partly because we keep running into bugs. Knowing whether this will be deprecated / can likely be stabilized would be great for planning!

Also, would this apply to all lambdas in general, or would you say that more controlled usage, i.e. explicit tracking with `track_on` etc., should, at least, be "fine"?

**Reply by zzzeek** [reply]
I found a fix which will be in the next 2.0.x release.   but overall im generally not very confident about this code

**Comment by CaselIT** [comment]
this doesn't do what you think:
```py
for crt in criteria:
        stmt += lambda s: s.where(crt)
```

example:
```pt
In [4]: f = [lambda: print(i) for i in range(4)]

In [5]: for l in f:
   ...:     l()
   ...:
```
this prints
```
3
3
3
3
```

**Reply by zzzeek** [reply]
it's still a bug if you look at what I have at #12084

i can't fix it without re-understanding this whole thing

**Reply by provinzkraut** [reply]
@CaselIT Good spot, but unfortunately that's just an oversight in my repro. Change it to:

```python
def append_to_lambda(stmt, crt):
    stmt += lambda s: s.where(crt)
    return stmt


def make_query(stmt, *criteria):
    for crt in criteria:
        stmt = append_to_lambda(stmt, crt)

    return stmt
```

and the behaviour is still the same.
Unless I'm misunderstanding lambda scoping in some way (not ruling that out), this should work.
